apiserver/oauth-proxy-proxy.md

```diff
@@ -0,0 +1,141 @@
+---
+title: OAuth-Proxy-Proxy-Support
+authors:
+  - "@deads2k"
+reviewers:
+  - "@sur"
+  - "@bparees"
+  - "@enj"
+  - "@sttts"
+approvers:
+  - "@sttts"
+creation-date: 2019-09-18
+last-updated: 2019-09-18
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# OAuth Proxy Proxy Support
+
+We should create zero-config method of enabling proxy-support for the oauth-proxy to enable non-expert teams to be secure
+when proxies are configured in a cluster.
+The `oauth-proxy` is used to front many webservers to provide authentication and authorization protection without directly
+building it into their binaries. This is common, easy, and used by people without a deep understand of how OpenShift is
+wired together.  They are not likely to know how to reactively wire proxy configuration and ca-bundles into their 
+side-cars.  Because it is re-used by directly embedding the `oauth-proxy` into a `PodSpec`, any changes to the way it is
+invoked will cause adoption difficulties.
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+In order to function, the `oauth-proxy`
+ 1. Gets the oauth-server location using the kube-apiserver's internal IP: `https://172.30.0.1/.well-known/oauth-authorization-server`
+ 2. Uses the result to find the token endpoint.
+ 3. Hits the *external* token endpoint `https://oauth-openshift.apps.<cluster-dns-prefix>.openshift.com/oauth/token`.
+
+This flow requires hitting a potentially external router, which requires honoring proxy settings.
+
+## Motivation
+
+See summary in first section.
+
+### Goals
+
+1. Zero-config proxy support that dynamically updates as proxy settings change.
+2. Support for all proxy modes, including MITM.
+
+### Non-Goals
+
+1. Making generic code to rewrite proxy transport configuration.
+
+## Proposal
+
+The `oauth-proxy` can be written to use targeted list/watch against the proxy.config.openshift.io and 
```

Comment:
Who will get access to `proxy.config.openshift.io`?

Reply:
What happens what a normal, unprivileged user, runs this new version of OAuth proxy.  It is used by people to add a simple layer of auth to various apps (usually those that interact with the kube API in some way).

Reply:
> What happens what a normal, unprivileged user, runs this new version of OAuth proxy. It is used by people to add a simple layer of auth to various apps (usually those that interact with the kube API in some way).

I think we have to expose this to system:authenticated.

@danehans @bparees this would mean that basic-auth will face challenges.

Reply:
> What happens what a normal, unprivileged user, runs this new version of OAuth proxy. It is used by people to add a simple layer of auth to various apps (usually those that interact with the kube API in some way).

I think we need to make it system:authenticated.
@danehans @bparees see this for basic-auth issues.